agent/eventhandler/attachment_handler.go

```diff
@@ -0,0 +1,150 @@
+}
+
+// NewAttachmentEventHandler returns a new AttachmentEventHandler object
+func NewAttachmentEventHandler(ctx context.Context,
+	stateSaver statemanager.Saver,
+	client api.ECSClient) *AttachmentEventHandler {
+	return &AttachmentEventHandler{
+		ctx:                    ctx,
+		stateSaver:             stateSaver,
+		client:                 client,
+		attachmentARNToHandler: make(map[string]*attachmentHandler),
+	}
+}
+
+// AddStateChangeEvent adds a state change event to AttachmentEventHandler for it to handle
+func (handler *AttachmentEventHandler) AddStateChangeEvent(change statechange.Event) error {
+	if change.GetEventType() != statechange.AttachmentEvent {
+		return fmt.Errorf("eventhandler: attachment handler received unrecognized event type: %d", change.GetEventType())
+	}
+	event, ok := change.(api.AttachmentStateChange)
+	if !ok {
+		return errors.New("eventhandler: unable to get attachment event from state change event")
+	}
+
+	if event.Attachment == nil {
+		return fmt.Errorf("eventhandler: received malformed attachment state change event: %v", event)
+	}
+
+	attachmentARN := event.Attachment.AttachmentARN
+	handler.lock.Lock()
+	if _, ok := handler.attachmentARNToHandler[attachmentARN]; !ok {
+		handler.attachmentARNToHandler[attachmentARN] = &attachmentHandler{
+			attachmentARN: attachmentARN,
+			stateSaver:    handler.stateSaver,
+			client:        handler.client,
+			ctx:           handler.ctx,
+		}
+	}
+	handler.lock.Unlock()
+
+	attachmentHandler := handler.attachmentARNToHandler[attachmentARN]
+	go attachmentHandler.submitAttachmentEvent(&event)
+
+	return nil
+}
+
+// submitAttachmentEvent submits an attachment event to backend
+func (handler *attachmentHandler) submitAttachmentEvent(attachmentChange *api.AttachmentStateChange) {
+	seelog.Debugf("AttachmentHandler: acquiring attachment lock before sending attachment state change for attachment %s", handler.attachmentARN)
+	handler.lock.Lock()
+	seelog.Debugf("AttachmentHandler: acquired attachment lock for attachment %s", handler.attachmentARN)
+	defer handler.lock.Unlock()
+
+	backoff := GetSubmitAttachmentStateBackoffFunc()
+
+	retry.RetryWithBackoffCtx(handler.ctx, backoff, func() error {
+		if !attachmentChangeShouldBeSent(attachmentChange) {
+			seelog.Debugf("AttachmentHandler: not sending attachment state change [%s] as it should not be sent", attachmentChange.String())
+			// if the attachment state change should not be sent, we don't need to retry anymore so return nil here
+			return nil
+		}
+
+		seelog.Infof("AttachmentHandler: sending attachment state change: %s", attachmentChange.String())
+		if err := handler.client.SubmitAttachmentStateChange(*attachmentChange); err != nil {
+			seelog.Errorf("AttachmentHandler: error submitting attachment state change [%s]: %v", attachmentChange.String(), err)
+			return err
+		}
+
+		attachmentChange.Attachment.SetSentStatus()
+		attachmentChange.Attachment.StopAckTimer()
+		handler.stateSaver.Save()
```

Comment:
Doesn't this method return an error? Shouldn't that be handled as well?

Reply:
added logic to handle state saving error here and added a unit test for it